storage/redis/peer_store.go

```diff
@@ -0,0 +1,787 @@
+	default:
+	}
+
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih // key
+	encoded_ih_S := IPver + "_S_" + encoded_ih // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechers, err := conn.Do("HKEYS", encoded_ih_L)
+	if err != nil {
+		return nil, err
+	}
+	con_leechers := leechers.([]interface{})
+
+	seeders, err := conn.Do("HKEYS", encoded_ih_S)
+	if err != nil {
+		return nil, err
+	}
+	con_seeders := seeders.([]interface{})
+
+	if len(con_leechers) == 0 && len(con_seeders) == 0 {
+		return nil, storage.ErrResourceDoesNotExist
+	}
+
+	if seeder {
+		// Append leechers as possible.
+		for _, pk := range con_leechers {
+			if numWant == 0 {
+				break
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+			numWant--
+		}
+	} else {
+		// Append as many seeders as possible.
+		for _, pk := range con_seeders {
+			if numWant == 0 {
+				break
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+			numWant--
+		}
+
+		// Append leechers until we reach numWant.
+		if numWant > 0 {
+			announcerPK := newPeerKey(announcer)
+			for _, pk := range con_leechers {
+				if pk == announcerPK {
+					continue
+				}
+
+				if numWant == 0 {
+					break
+				}
+
+				peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+				numWant--
+			}
+		}
+	}
+
+	APResult := ""
+	for _, pr := range peers {
+		APResult = fmt.Sprintf("%s Peer:[ID: %s, IP: %s(AddressFamily: %s), Port %d]", APResult, pr.ID.String(), pr.IP.String(), IPver, pr.Port)
+	}
+	log.Debug("storage: AnnouncePeers result", log.Fields{
+		"peers": APResult,
+	})
+
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	resp.InfoHash = ih
+	IPver := addressFamily.String()
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih // key
+	encoded_ih_S := IPver + "_S_" + encoded_ih // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechers_len, err := conn.Do("HLEN", encoded_ih_L)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encoded_ih_L,
+			"error": err,
+		})
+		return
+	}
+	l_len := leechers_len.(int64)
+
+	seeders_len, err := conn.Do("HLEN", encoded_ih_S)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encoded_ih_S,
+			"error": err,
+		})
+		return
+	}
+	s_len := seeders_len.(int64)
+
+	if l_len == 0 && s_len == 0 {
+		return
+	}
+
+	resp.Incomplete = uint32(l_len)
+	resp.Complete = uint32(s_len)
+
+	return
+}
+
+// collectGarbage deletes all Peers from the PeerStore which are older than the
+// cutoff time.
+//
+// This function must be able to execute while other methods on this interface
+// are being executed in parallel.
+func (ps *peerStore) collectGarbage(cutoff time.Time) error {
```

Comment:
Redis has some concept of key expiry, right? I'm not an expert about redis, but that would've been my first choice - the garbage collection is just used for the memory implementation because Go doesn't provide kex expiry. Why should we/should we not use the redis feature?

Reply:
Redis can't expire within a hash. A hash makes sense here to provide performant iteration, though a hash will also have issues with many keys when there are many peers.

Reply:
Oh, that's unfortunate. Thanks for clearing that up!

Alternative idea: Implement the whole garbage collectio thing as a lua script? Although that would completely stop-the-world while it's running, right?

Reply:
@mrd0ll4r @pushrax Maybe it is unnecessary to do synchronization during garbage collection as `lost one or serveral peers` during this interval doesn't affect the overall Peers conditions(there are still many other peers we can acquire).

Reply:
That's true, yes. Off the top of my head, the only "important" requirement would be that inactive peers/swarms are cleaned up eventually. Opinions @pushrax @jzelinskie ?